File: packages/php/src/Resources/Emails.php
<?php

declare(strict_types=1);

namespace Axene\Mailer\Resources;

use Axene\Mailer\Internal;
use Axene\Mailer\Transport;

/**
 * The `emails` resource: send, look up, search, schedule, and inspect messages.
 * Accessed as `$client->emails`.
 */
final class Emails
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * Send a single email.
     *
     * The `from`, `to`, `cc`, `bcc`, and `replyTo` fields accept either an
     * address array (`['email' => ..., 'name' => ...]`) or a bare string,
     * which is sugar for `['email' => $string]`.
     *
     * @param array<string, mixed> $params
     * @return array<string, mixed>
     */
    public function send(array $params): array
    {
        return $this->transport->request('POST', '/v1/emails/', Internal::serializeSend($params));
    }

    /**
     * Send up to your plan's batch limit in one call. The API accepts a bare
     * array of messages and returns a per-message result set.
     *
     * @param list<array<string, mixed>> $messages
     * @return array<string, mixed>
     */
    public function sendBatch(array $messages): array
    {
        $body = array_map(static fn (array $m) => Internal::serializeSend($m), $messages);

        return $this->transport->request('POST', '/v1/emails/batch', $body);
    }

    /**
     * Dry-run a send: check whether `message` would be accepted (sender
     * registered, domain verified, plan limits) without actually sending it.
     *
     * @param array<string, mixed> $message
     * @return array<string, mixed>
     */
    public function validate(array $message): array
    {
        return $this->transport->request('POST', '/v1/emails/validate', Internal::serializeSend($message));
    }

    /**
     * List recent emails, newest first. Zero-based `page`.
     *
     * @param array{status?: string, page?: int, limit?: int} $params
     * @return list<array<string, mixed>>
     */
    public function list(array $params = []): array
    {
        return $this->transport->request('GET', '/v1/emails/' . Internal::query($params));
    }

    /**
     * Fetch a single email with its bodies and events.
     *
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return $this->transport->request('GET', '/v1/emails/' . rawurlencode($id));
    }

    /**
     * List delivery / open / click / bounce events for an email.
     *
     * @return list<array<string, mixed>>
     */
    public function events(string $id): array
    {
        return $this->transport->request('GET', '/v1/emails/' . rawurlencode($id) . '/events');
    }

    /**
     * Re-send a bounced, rejected, or failed email as a new message.
     *
     * @return array<string, mixed>
     */
    public function retry(string $id): array
    {
        return $this->transport->request('POST', '/v1/emails/' . rawurlencode($id) . '/retry');
    }

    /**
     * Search emails. `q` supports inline tokens (`to:`, `from:`, `status:`,
     * `domain:`, `tag:`); leftover words are matched as free text.
     *
     * @param array{q?: string, status?: string, tag?: string, page?: int, limit?: int} $params
     * @return list<array<string, mixed>>
     */
    public function search(array $params = []): array
    {
        return $this->transport->request('GET', '/v1/emails/search' . Internal::query($params));
    }

    /**
     * List emails scheduled for future delivery, soonest first.
     *
     * @return list<array<string, mixed>>
     */
    public function listScheduled(): array
    {
        return $this->transport->request('GET', '/v1/emails/scheduled');
    }

    /**
     * Cancel a scheduled email.
     *
     * @return array<string, mixed>
     */
    public function cancelScheduled(string $id): array
    {
        return $this->transport->request('DELETE', '/v1/emails/scheduled/' . rawurlencode($id));
    }

    /**
     * Send a scheduled email immediately instead of waiting.
     *
     * @return array<string, mixed>
     */
    public function sendScheduledNow(string $id): array
    {
        return $this->transport->request('POST', '/v1/emails/scheduled/' . rawurlencode($id) . '/send-now');
    }

    /**
     * Poll for emails whose status changed at or after `since` (ISO 8601).
     * Capped at 50 rows; use for live status updates.
     *
     * @return list<array<string, mixed>>
     */
    public function updates(string $since): array
    {
        return $this->transport->request('GET', '/v1/emails/updates' . Internal::query(['since' => $since]));
    }

    /**
     * Get the caller's saved searches.
     *
     * @return list<array<string, mixed>>
     */
    public function getSavedSearches(): array
    {
        $response = $this->transport->request('GET', '/v1/emails/saved-searches');

        return $response['searches'] ?? [];
    }

    /**
     * Replace the caller's saved searches (max 50).
     *
     * @param list<array<string, mixed>> $searches
     * @return list<array<string, mixed>>
     */
    public function setSavedSearches(array $searches): array
    {
        $response = $this->transport->request('PUT', '/v1/emails/saved-searches', ['searches' => $searches]);

        return $response['searches'] ?? [];
    }
}

File: packages/rust/src/resources/emails.rs
//! The `emails` resource: send, look up, search, schedule, and inspect messages.

use reqwest::Method;
use serde::Deserialize;
use serde_json::Value;

use super::{query, urlencode};
use crate::error::Result;
use crate::http::HttpTransport;
use crate::models::{
    BatchResponse, Email, EmailDetail, EmailEvent, EmailSearchHit, IdStatus, ScheduledEmail,
    SendEmail, SendEmailResponse, ValidationResult,
};

/// Accessed as `client.emails()`.
#[derive(Debug, Clone)]
pub struct Emails {
    http: HttpTransport,
}

#[derive(Deserialize)]
struct SavedSearches {
    searches: Vec<Value>,
}

impl Emails {
    pub(crate) fn new(http: HttpTransport) -> Self {
        Self { http }
    }

    /// Send a single email.
    pub async fn send(&self, message: &SendEmail) -> Result<SendEmailResponse> {
        self.http
            .request_json(Method::POST, "/v1/emails/", message)
            .await
    }

    /// Send up to your plan's batch limit in one call. The API accepts a bare
    /// array of messages and returns a per-message result set.
    pub async fn send_batch(&self, messages: &[SendEmail]) -> Result<BatchResponse> {
        self.http
            .request_json(Method::POST, "/v1/emails/batch", &messages)
            .await
    }

    /// Dry-run a send: check whether the message would be accepted without
    /// actually sending it.
    pub async fn validate(&self, message: &SendEmail) -> Result<ValidationResult> {
        self.http
            .request_json(Method::POST, "/v1/emails/validate", message)
            .await
    }

    /// List recent emails, newest first. `page` is zero-based.
    pub async fn list(
        &self,
        status: Option<&str>,
        page: Option<u64>,
        limit: Option<u64>,
    ) -> Result<Vec<Email>> {
        let q = query(&[
            ("status", status.map(String::from)),
            ("page", page.map(|p| p.to_string())),
            ("limit", limit.map(|l| l.to_string())),
        ]);
        self.http
            .request(Method::GET, &format!("/v1/emails/{q}"))
            .await
    }

    /// Fetch a single email with its bodies and events.
    pub async fn get(&self, id: &str) -> Result<EmailDetail> {
        self.http
            .request(Method::GET, &format!("/v1/emails/{}", urlencode(id)))
            .await
    }

    /// List delivery / open / click / bounce events for an email.
    pub async fn events(&self, id: &str) -> Result<Vec<EmailEvent>> {
        self.http
            .request(
                Method::GET,
                &format!("/v1/emails/{}/events", urlencode(id)),
            )
            .await
    }

    /// Re-send a bounced, rejected, or failed email as a new message.
    pub async fn retry(&self, id: &str) -> Result<SendEmailResponse> {
        self.http
            .request(Method::POST, &format!("/v1/emails/{}/retry", urlencode(id)))
            .await
    }

    /// Search emails. `q` supports inline tokens (`to:`, `from:`, `status:`,
    /// `domain:`, `tag:`); leftover words are matched as free text.
    pub async fn search(
        &self,
        q: Option<&str>,
        status: Option<&str>,
        tag: Option<&str>,
        page: Option<u64>,
        limit: Option<u64>,
    ) -> Result<Vec<EmailSearchHit>> {
        let qs = query(&[
            ("q", q.map(String::from)),
            ("status", status.map(String::from)),
            ("tag", tag.map(String::from)),
            ("page", page.map(|p| p.to_string())),
            ("limit", limit.map(|l| l.to_string())),
        ]);
        self.http
            .request(Method::GET, &format!("/v1/emails/search{qs}"))
            .await
    }

    /// List emails scheduled for future delivery, soonest first.
    pub async fn list_scheduled(&self) -> Result<Vec<ScheduledEmail>> {
        self.http
            .request(Method::GET, "/v1/emails/scheduled")
            .await
    }

    /// Cancel a scheduled email.
    pub async fn cancel_scheduled(&self, id: &str) -> Result<IdStatus> {
        self.http
            .request(
                Method::DELETE,
                &format!("/v1/emails/scheduled/{}", urlencode(id)),
            )
            .await
    }

    /// Send a scheduled email immediately instead of waiting.
    pub async fn send_scheduled_now(&self, id: &str) -> Result<IdStatus> {
        self.http
            .request(
                Method::POST,
                &format!("/v1/emails/scheduled/{}/send-now", urlencode(id)),
            )
            .await
    }

    /// Poll for emails whose status changed at or after `since` (ISO 8601).
    /// Capped at 50 rows.
    pub async fn updates(&self, since: &str) -> Result<Vec<Email>> {
        let q = query(&[("since", Some(since.to_string()))]);
        self.http
            .request(Method::GET, &format!("/v1/emails/updates{q}"))
            .await
    }

    /// Get the caller's saved searches.
    pub async fn get_saved_searches(&self) -> Result<Vec<Value>> {
        let r: SavedSearches = self
            .http
            .request(Method::GET, "/v1/emails/saved-searches")
            .await?;
        Ok(r.searches)
    }

    /// Replace the caller's saved searches (max 50).
    pub async fn set_saved_searches(&self, searches: Vec<Value>) -> Result<Vec<Value>> {
        let body = serde_json::json!({ "searches": searches });
        let r: SavedSearches = self
            .http
            .request_json(Method::PUT, "/v1/emails/saved-searches", &body)
            .await?;
        Ok(r.searches)
    }
}
